src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+  DOSBox integration by kekko (https://www.vogons.org/viewtopic.php?f=41&t=41853)
+  Code cleanups and multi-threaded triangle rendering by Bernhard Schelling
+*/
+
+#include "dosbox.h"
+
+#include <assert.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <SDL.h>
+
+#if C_VOODOO
+#include "control.h"
+#include "cross.h"
+#include "mem.h"
+#include "paging.h"
+#include "pci_bus.h"
+#include "pic.h"
+#include "render.h"
+#include "setup.h"
+#include "vga.h"
+
+#ifndef DOSBOX_VOODOO_TYPES_H
+#define DOSBOX_VOODOO_TYPES_H
+
+/***************************************************************************
+    TYPE DEFINITIONS
+***************************************************************************/
+
+// FIXME Get rid of these typedefs
+//
+/* 8-bit values */
+typedef uint8_t UINT8;
+typedef int8_t INT8;
+
+/* 16-bit values */
+typedef uint16_t UINT16;
+typedef int16_t INT16;
+
+#ifndef _WINDOWS_
+/* 32-bit values */
+typedef uint32_t UINT32;
+typedef int32_t INT32;
+
+/* 64-bit values */
+typedef uint64_t UINT64;
+typedef int64_t INT64;
+#endif
+
+typedef INT64 attoseconds_t;
+
+#define ATTOSECONDS_PER_SECOND_SQRT		((attoseconds_t)1000000000)
+#define ATTOSECONDS_PER_SECOND			(ATTOSECONDS_PER_SECOND_SQRT * ATTOSECONDS_PER_SECOND_SQRT)
+
+/* convert between hertz (as a double) and attoseconds */
+#define ATTOSECONDS_TO_HZ(x)			((double)ATTOSECONDS_PER_SECOND / (double)(x))
+#define HZ_TO_ATTOSECONDS(x)			((attoseconds_t)(ATTOSECONDS_PER_SECOND / (x)))
+
+#define MAX_VERTEX_PARAMS					6
+
+/* poly_extent describes start/end points for a scanline, along with per-scanline parameters */
+struct poly_extent
+{
+	INT32		startx;						/* starting X coordinate (inclusive) */
+	INT32		stopx;						/* ending X coordinate (exclusive) */
+};
+
+/* an rgb_t is a single combined R,G,B (and optionally alpha) value */
+typedef UINT32 rgb_t;
+
+/* an rgb15_t is a single combined 15-bit R,G,B value */
+typedef UINT16 rgb15_t;
+
+/* macros to assemble rgb_t values */
+#define MAKE_ARGB(a,r,g,b)	((((rgb_t)(a) & 0xff) << 24) | (((rgb_t)(r) & 0xff) << 16) | (((rgb_t)(g) & 0xff) << 8) | ((rgb_t)(b) & 0xff))
+#define MAKE_RGB(r,g,b)		(MAKE_ARGB(255,r,g,b))
+
+/* macros to extract components from rgb_t values */
+#define RGB_ALPHA(rgb)		(((rgb) >> 24) & 0xff)
+#define RGB_RED(rgb)		(((rgb) >> 16) & 0xff)
+#define RGB_GREEN(rgb)		(((rgb) >> 8) & 0xff)
+#define RGB_BLUE(rgb)		((rgb) & 0xff)
+
+/* common colors */
+#define RGB_BLACK			(MAKE_ARGB(255,0,0,0))
+#define RGB_WHITE			(MAKE_ARGB(255,255,255,255))
+
+/***************************************************************************
+    inline FUNCTIONS
+***************************************************************************/
+
+/*-------------------------------------------------
+    pal5bit - convert a 5-bit value to 8 bits
+-------------------------------------------------*/
+inline UINT8 pal5bit(UINT8 bits)
+{
+	bits &= 0x1f;
+	return (bits << 3) | (bits >> 2);
+}
```

Comment:
Constexpr candidate perhaps? Seems like could be converted to a lookup as it's a small table?

Reply:
Agreed, to be fixed in future PR.